stream-chat-android-ui-common/src/main/kotlin/io/getstream/chat/android/ui/common/feature/channel/info/ChannelInfoController.kt

```diff
@@ -0,0 +1,420 @@
+                    logger.e { "[unhideChannel] error: ${error.message}" }
+                    _events.tryEmit(
+                        ChannelInfoEvent.UnhideChannelError(message = error.message),
+                    )
+                }
+        }
+    }
+
+    /**
+     * Leaves the channel with an optional quit message.
+     *
+     * @param quitMessage The system message to send when leaving the channel.
+     */
+    public fun leaveChannel(quitMessage: Message?) {
+        logger.d { "[leaveChannel] quitMessage: ${quitMessage?.text}" }
+
+        val onError: (Error) -> Unit = { error ->
+            logger.e { "[leaveChannel] error: ${error.message}" }
+            _events.tryEmit(
+                ChannelInfoEvent.LeaveChannelError(message = error.message),
+            )
+        }
+
+        requireCapability(
+            permission = { canLeaveChannel },
+            onError = onError,
+        ) {
+            scope.launch {
+                runCatching {
+                    requireNotNull(chatClient.getCurrentOrStoredUserId()) { "User not connected" }
+                }.onSuccess { currentUserId ->
+                    channelClient.removeMembers(
+                        memberIds = listOf(currentUserId),
+                        systemMessage = quitMessage,
+                    ).await()
+                        .onSuccess { _events.tryEmit(ChannelInfoEvent.LeaveChannelSuccess) }
+                        .onError(onError)
+                }.onFailure { cause ->
+                    onError(Error.ThrowableError(message = cause.message.orEmpty(), cause = cause))
+                }
+            }
+        }
+    }
+
+    /**
+     * Deletes the channel.
+     */
+    public fun deleteChannel() {
+        logger.d { "[deleteChannel]" }
+
+        val onError: (Error) -> Unit = { error ->
+            logger.e { "[deleteChannel] error: ${error.message}" }
+            _events.tryEmit(
+                ChannelInfoEvent.DeleteChannelError(message = error.message),
+            )
+        }
+
+        requireCapability(
+            permission = { canDeleteChannel },
+            onError = onError,
+        ) {
+            scope.launch {
+                channelClient.delete().await()
+                    .onSuccess { _events.tryEmit(ChannelInfoEvent.DeleteChannelSuccess) }
+                    .onError(onError)
+            }
+        }
+    }
+
+    private fun List<Member>.filterNotCurrentUser() =
+        filter { member -> member.user.id != chatClient.getCurrentOrStoredUserId() }
+
+    private fun requireCapability(
+        permission: ChannelInfoState.Content.Capability.() -> Boolean,
+        onError: (error: Error) -> Unit,
+        onSuccess: () -> Unit,
+    ) {
+        (_state.value as? ChannelInfoState.Content)?.capability?.let { capability ->
+            runCatching { require(permission(capability)) }
+                .onFailure { onError(Error.GenericError("User doesn't have permission")) }
```

Comment:
Can't we forward the previous error?
It is hiding the root cause of the failure

Reply:
`permission` is a `ChannelInfoState.Content.Capability.() -> Boolean`
`require` is a Kotlin function which throw if the parameter is not true.
The thrown message is too generic: `Failed requirement`
The actual cause is due to the user not having permission enabled